File: shared/src/protocol.ts
import { State, PermissionMode, PromptOption } from './states.js';
import type { AgentType, AgentCapabilities } from './adapter.js';
import type { TimelineEntry } from './timeline.js';

// ===== Billing Type =====

export type BillingType = 'subscription' | 'api' | 'unknown';

// ===== Model Catalog (OpenClaw) =====

export interface ModelCatalogEntry {
  key: string;
  name: string;
  role: 'default' | `fallback-${number}` | 'configured';
  available: boolean;
}

// ===== OpenClaw Session Status =====

export interface OcSessionStatus {
  model?: string;
  contextTokens?: number;
  messageCount?: number;
  uptime?: string;
  sessionId?: string;
  [key: string]: unknown;
}

// ===== Button State (Bridge → Android) =====

export interface ButtonSlotState {
  slot: number;           // 0-7
  title: string;
  subtitle?: string;
  bgColor: string;        // hex
  textColor: string;      // hex
  enabled: boolean;
  icon?: string;
  badge?: string;         // "★", "1/3"
  action?: string;        // "switch_mode" | "command:go on" | "respond:y" | "select_option:2" | "interrupt" | "escape" | "expand_options"
  dim?: boolean;
}

export interface ButtonStateEvent {
  type: 'button_state';
  buttons: ButtonSlotState[];
}

// ===== Encoder LCD State (Bridge → Plugin/Android) =====

export interface EncoderSlotState {
  slot: number; // 0-3 (E1~E4)
  encoderType: 'utility' | 'action' | 'usage' | 'voice';
  header: string;         // "VOLUME", "PROMPT", "SESSION", "VOICE"
  value?: string;         // "65%", "go on", session name, "Ready"
  icon?: string;          // emoji
  accentColor: string;    // hex — bottom indicator bar color
  progress?: number;      // 0-1, indicator bar fill
  counter?: string;       // "1/4" format
  detail?: string;        // additional text (option description, error msg, etc.)
  // Voice specific
  voiceState?: 'idle' | 'recording' | 'transcribing' | 'error' | 'review';
  recordingMs?: number;
  transcription?: string;
}

export interface EncoderStateEvent {
  type: 'encoder_state';
  encoders: EncoderSlotState[];
  takeoverActive: boolean; // AWAITING states: all encoders show options
}

// ===== Deck Slot Map (Plugin → Bridge → Android) =====

export interface DeckSlotConfig {
  slot: number;
  actionType: string; // 'mode-button' | 'session-button' | 'usage-button' | 'response-button' | 'stop-button' | 'utility-dial' | 'option-dial' | 'iterm-dial' | 'voice-dial'
  settings?: Record<string, unknown>;
}

export interface DeckSlotMapEvent {
  type: 'deck_slot_map';
  buttons: DeckSlotConfig[];
  encoders: DeckSlotConfig[];
}

// ===== Ollama Status =====

export interface OllamaModel {
  name: string;
  size: number;
  sizeVram: number;
}

export interface OllamaStatus {
  available: boolean;
  models: OllamaModel[];
}

export interface SubscriptionInfo {
  name: string;
  until?: string;
}

export interface AntigravityStatusInfo {
  planName?: string;
  availableCredits?: number;
  minimumCreditAmountForUsage?: number;
}

// ===== Bridge → Plugin (State Updates) =====

export interface StateUpdateEvent {
  type: 'state_update';
  state: State;
  permissionMode: PermissionMode;
  agentType?: AgentType;
  agentCapabilities?: AgentCapabilities;
  currentTool?: string;
  toolInput?: string;
  toolProgress?: string;
  projectName?: string;
  modelName?: string;
  effortLevel?: string;
  billingType?: BillingType;
  options?: PromptOption[];
  promptType?: 'yes_no' | 'yes_no_always' | 'multi_select' | 'diff_review';
  question?: string;
  navigable?: boolean;
  cursorIndex?: number;
  suggestedPrompt?: string;
  modelCatalog?: ModelCatalogEntry[];
  sessionStatus?: OcSessionStatus;
  remoteUrl?: string;
  /** Authenticated WS URL for remote pairing (ws://ip:port?token=hex) */
  pairingUrl?: string;
  /** Number of OpenClaw backend worker sessions (multi-agent) */
  workerSessionCount?: number;
  /** Ollama process status + running models */
  ollamaStatus?: OllamaStatus;
  /** MLX local server model list */
  mlxModels?: string[];
  /** Subscription-backed authenticated services */
  subscriptions?: SubscriptionInfo[];
  /** Local Antigravity IDE quota summary, when available */
  antigravityStatus?: AntigravityStatusInfo;
  /** OpenClaw Gateway reachability (port 18789) */
  gatewayAvailable?: boolean;
  /** OpenClaw Gateway has doctor warnings/errors */
  gatewayHasError?: boolean;
  /** Voice assistant pipeline state (wake word → STT → LLM → TTS) */
  voiceAssistantState?: VoiceAssistantState;
  /** Transcribed user speech (processing/speaking) */
  voiceAssistantText?: string;
  /** LLM response text (speaking) */
  voiceAssistantResponseText?: string;
}

export interface PromptOptionsEvent {
  type: 'prompt_options';
  promptType: 'yes_no' | 'yes_no_always' | 'multi_select' | 'diff_review';
  question?: string;
  options: PromptOption[];
}

export interface UsageEvent {
  type: 'usage_update';
  sessionDurationSec: number;
  inputTokens: number;
  outputTokens: number;
  toolCalls: number;
  estimatedCostUsd?: number;
  // Legacy PTY-parsed fields (kept for backward compat)
  sessionPercent?: number;
  costSpent?: number;
  costLimit?: number;
  resetTime?: string;
  resetDate?: string;
  // API-sourced global usage
  fiveHourPercent?: number;
  fiveHourResetsAt?: string;
  sevenDayPercent?: number;
  sevenDayResetsAt?: string;
  // Extra usage (pay-per-use beyond plan limits)
  extraUsageEnabled?: boolean;
  extraUsageMonthlyLimit?: number;
  extraUsageUsedCredits?: number;
  extraUsageUtilization?: number;
  // OAuth connection status
  oauthConnected?: boolean;
  // Ollama process status + running models (piggyback on usage polling)
  ollamaStatus?: OllamaStatus;
  // True when displaying cached data after a fetch failure
  usageStale?: boolean;
  // OAuth token status: valid/expired/missing/unknown
  tokenStatus?: 'valid' | 'expired' | 'missing' | 'unknown';
  // Codex / ChatGPT web-auth metadata
  codexAuthMode?: string;
  codexWebAuthConnected?: boolean;
  codexPlanType?: string;
  codexAccountId?: string;
  codexSubscriptionActiveUntil?: string;
  codexLastRefreshAt?: string;
  // Local model/runtime summaries
  modelCatalog?: ModelCatalogEntry[];
  mlxModels?: string[];
  subscriptions?: SubscriptionInfo[];
  antigravityStatus?: AntigravityStatusInfo;
}

export interface ConnectionEvent {
  type: 'connection';
  status: 'connected' | 'reconnecting' | 'disconnected';
  sessionId?: string;
}

export interface UserPromptEvent {
  type: 'user_prompt';
  text: string;
}

export interface VoiceStateEvent {
  type: 'voice_state';
  state: 'idle' | 'recording' | 'transcribing' | 'error';
  text?: string;
  error?: string;
}

// ===== Voice Assistant (Wake Word) =====

export type VoiceAssistantState =
  | 'idle'        // listening for wake word
  | 'listening'   // wake word detected, recording user speech
  | 'processing'  // STT + LLM in progress
  | 'speaking'    // TTS playback
  | 'disabled';   // wake word listener off

export interface VoiceAssistantStateEvent {
  type: 'voice_assistant_state';
  state: VoiceAssistantState;
  deviceId?: string;
  /** Transcribed user speech */
  text?: string;
  /** LLM response text */
  responseText?: string;
}

export interface WakeWordDetectedEvent {
  type: 'wake_word_detected';
  deviceId: string;
  timestamp: number;
}

export interface DisplayStateEvent {
  type: 'display_state';
  displayOn: boolean;
}

// ===== Multi-session Discovery =====

export interface SessionInfo {
  id: string;
  port: number;
  projectName: string;
  agentType?: AgentType;
  alive: boolean;
  state?: string;  // sibling's current state from /health query
  modelName?: string;  // sibling's current model from /health query
}

export interface SessionsListEvent {
  type: 'sessions_list';
  sessions: SessionInfo[];
}

export interface TimelineEventMsg {
  type: 'timeline_event';
  entry: TimelineEntry;
  upsert?: boolean;
}

export interface TimelineHistoryMsg {
  type: 'timeline_history';
  entries: TimelineEntry[];
}

// ===== WiFi Provisioning (Bridge → ESP32 Serial) =====

export interface WifiProvisionMessage {
  type: 'wifi_provision';
  ssid: string;
  password: string;
  bridgeIp: string;
  bridgePort: number;
  authToken: string;
}

// ===== ESP32 Serial → Bridge =====

export interface DeviceInfoMessage {
  type: 'device_info';
  board: string;         // "86box" | "round_amoled" | "ips_35"
  version: string;       // firmware version
  wifiConfigured: boolean;
  wifiConnected: boolean;
  ip?: string;
}

export interface WifiProvisionAckMessage {
  type: 'wifi_provision_ack';
  success: boolean;
  ip?: string;           // assigned IP on success
  error?: string;        // reason on failure
}

export interface WifiStatusMessage {
  type: 'wifi_status';
  connected: boolean;
  ssid?: string;
  ip?: string;
}

export type ESP32ToHostMessage =
  | DeviceInfoMessage
  | WifiProvisionAckMessage
  | WifiStatusMessage;

export type BridgeEvent =
  | StateUpdateEvent
  | PromptOptionsEvent
  | UsageEvent
  | ConnectionEvent
  | UserPromptEvent
  | VoiceStateEvent
  | VoiceAssistantStateEvent
  | WakeWordDetectedEvent
  | DisplayStateEvent
  | SessionsListEvent
  | EncoderStateEvent
  | DeckSlotMapEvent
  | ButtonStateEvent
  | TimelineEventMsg
  | TimelineHistoryMsg;

// ===== Plugin → Bridge (Commands) =====

export interface ResponseCommand {
  type: 'respond';
  value: string; // 'y' | 'n' | 'a' or option text
}

export interface SelectOptionCommand {
  type: 'select_option';
  index: number; // 0-based option index
}

export interface NavigateOptionCommand {
  type: 'navigate_option';
  direction: 'up' | 'down';
}

export interface SendPromptCommand {
  type: 'send_prompt';
  text: string;
}

export interface SwitchModeCommand {
  type: 'switch_mode';
  mode?: 'plan' | 'acceptEdits' | 'default';
}

export interface InterruptCommand {
  type: 'interrupt'; // Ctrl+C
}

export interface EscapeCommand {
  type: 'escape'; // Esc key — cancel prompt/selection
}

export interface VoiceCommand {
  type: 'voice';
  action: 'start' | 'stop' | 'cancel';
}

export interface QueryUsageCommand {
  type: 'query_usage';
}

export interface DiagCommand {
  type: 'diag';
  action: 'dump' | 'analyze';
}

export interface UtilityCommand {
  type: 'utility';
  action: 'adjust_volume' | 'toggle_mute' | 'adjust_brightness'
        | 'media_play_pause' | 'media_next' | 'media_prev';
  value?: number; // delta ticks (for adjust commands)
}

export interface SwitchAgentCommand {
  type: 'switch_agent';
  agent: 'openclaw' | 'claude-code';
}

export interface FocusSessionCommand {
  type: 'focus_session';
  sessionId: string;
}

/**
 * Session-scoped command — daemon forwards the inner command to the specified session's bridge.
 * Enables direct control of a specific session from any client (MenuBarExtra, Dashboard, etc.)
 */
export interface SessionCommand {
  type: 'session_command';
  sessionId: string;
  command: {
    type: string;
    [key: string]: unknown;
  };
}

export type PluginCommand =
  | ResponseCommand
  | SelectOptionCommand
  | NavigateOptionCommand
  | SendPromptCommand
  | SwitchModeCommand
  | InterruptCommand
  | EscapeCommand
  | VoiceCommand
  | QueryUsageCommand
  | DiagCommand
  | UtilityCommand
  | SwitchAgentCommand
  | FocusSessionCommand
  | SessionCommand;

// ===== Hook Event Types =====

export interface HookEvent {
  event: string;
  data: Record<string, unknown>;
}

// ===== Constants =====

/** Events forwarded to display-only devices (Pixoo64) */
export const DISPLAY_FORWARDED_EVENTS = new Set([
  'state_update',
  'usage_update',
  'sessions_list',
  'connection',
  'display_state',
]);

/** Events forwarded to serial devices (ESP32) — display events + timeline */
export const SERIAL_FORWARDED_EVENTS = new Set([
  ...DISPLAY_FORWARDED_EVENTS,
  'timeline_event',
  'timeline_history',
  'set_orientation',
]);

export const BRIDGE_WS_PORT = 9120;
export const BRIDGE_HTTP_PORT = 9120; // Same port, different path
export const RECONNECT_INTERVAL_MS = 3000;
export const STUCK_TIMEOUT_MS = 5 * 60 * 1000; // 5 minutes
export const WS_PING_INTERVAL_MS = 15_000;
export const WS_ACTIVITY_TIMEOUT_MS = 30_000;

File: android/app/src/main/kotlin/dev/agentdeck/net/Protocol.kt
package dev.agentdeck.net

import android.util.Log
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.jsonPrimitive

@Serializable
enum class AgentState {
    @SerialName("disconnected") DISCONNECTED,
    @SerialName("idle") IDLE,
    @SerialName("processing") PROCESSING,
    @SerialName("awaiting_permission") AWAITING_PERMISSION,
    @SerialName("awaiting_option") AWAITING_OPTION,
    @SerialName("awaiting_diff") AWAITING_DIFF,
}

@Serializable
enum class PermissionMode {
    @SerialName("default") DEFAULT,
    @SerialName("plan") PLAN,
    @SerialName("acceptEdits") ACCEPT_EDITS,
    @SerialName("dontAsk") DONT_ASK,
    @SerialName("bypassPermissions") BYPASS_PERMISSIONS,
}

@Serializable
data class AgentCapabilities(
    val type: String? = null,
    val displayName: String? = null,
    val hasTerminal: Boolean = false,
    val hasModeSwitching: Boolean = false,
    val hasDiffReview: Boolean = false,
    val hasOptionLists: Boolean = false,
    val hasNavigablePrompts: Boolean = false,
    val hasSuggestedPrompts: Boolean = false,
    val hasApiUsage: Boolean = false,
    val hasModelCatalog: Boolean = false,
)

@Serializable
data class ModelCatalogEntry(
    val name: String,
    val role: String? = null,
    val available: Boolean = true,
)

@Serializable
data class OcSessionStatus(
    val model: String? = null,
    val contextTokens: Int? = null,
    val messageCount: Int? = null,
    val uptime: Int? = null,
    val sessionId: String? = null,
)

@Serializable
data class PromptOption(
    val label: String,
    val value: String? = null,
    val description: String? = null,
    val index: Int? = null,
    val shortcut: String? = null,
    val recommended: Boolean? = null,
    val selected: Boolean? = null,
)

@Serializable
data class StateUpdate(
    val state: AgentState = AgentState.DISCONNECTED,
    val permissionMode: PermissionMode = PermissionMode.DEFAULT,
    val agentType: String? = null,
    val currentTool: String? = null,
    val toolInput: String? = null,
    val toolProgress: String? = null,
    val projectName: String? = null,
    val modelName: String? = null,
    val effortLevel: String? = null,
    val billingType: String? = null,
    val options: List<PromptOption>? = null,
    val promptType: String? = null,
    val question: String? = null,
    val suggestedPrompt: String? = null,
    val remoteUrl: String? = null,
    val navigable: Boolean? = null,
    val cursorIndex: Int? = null,
    val agentCapabilities: AgentCapabilities? = null,
    val modelCatalog: List<ModelCatalogEntry>? = null,
    val sessionStatus: OcSessionStatus? = null,
    val pairingUrl: String? = null,
    val workerSessionCount: Int? = null,
    val ollamaStatus: OllamaStatus? = null,
    val gatewayAvailable: Boolean? = null,
    val gatewayHasError: Boolean? = null,
)

@Serializable
data class UsageUpdate(
    val sessionDurationSec: Int = 0,
    val inputTokens: Int = 0,
    val outputTokens: Int = 0,
    val toolCalls: Int = 0,
    val estimatedCostUsd: Double? = null,
    val fiveHourPercent: Double? = null,
    val sevenDayPercent: Double? = null,
    val fiveHourResetsAt: String? = null,
    val sevenDayResetsAt: String? = null,
    val extraUsageEnabled: Boolean? = null,
    val extraUsageMonthlyLimit: Double? = null,
    val extraUsageUsedCredits: Double? = null,
    val extraUsageUtilization: Double? = null,
    val sessionPercent: Double? = null,
    val costSpent: Double? = null,
    val costLimit: Double? = null,
    val resetTime: String? = null,
    val resetDate: String? = null,
    val oauthConnected: Boolean? = null,
    val ollamaStatus: OllamaStatus? = null,
    val usageStale: Boolean? = null,
)

@Serializable
data class OllamaModel(
    val name: String,
    val size: Long = 0,
    val sizeVram: Long = 0,
)

@Serializable
data class OllamaStatus(
    val available: Boolean = false,
    val models: List<OllamaModel> = emptyList(),
)

@Serializable
data class VoiceState(
    val state: String = "idle",
    val text: String? = null,
    val error: String? = null,
)

// --- Button state (Bridge → Android) ---

@Serializable
data class ButtonSlotState(
    val slot: Int,
    val title: String,
    val subtitle: String? = null,
    val bgColor: String,
    val textColor: String,
    val enabled: Boolean = true,
    val icon: String? = null,
    val badge: String? = null,
    val action: String? = null,
    val dim: Boolean = false,
)

// --- Encoder LCD state (Bridge → Android) ---

@Serializable
data class EncoderSlotState(
    val slot: Int,
    val encoderType: String,
    val header: String,
    val value: String? = null,
    val icon: String? = null,
    val accentColor: String = "#94A3B8",
    val progress: Float? = null,
    val counter: String? = null,
    val detail: String? = null,
    val voiceState: String? = null,
    val recordingMs: Long? = null,
    val transcription: String? = null,
)

@Serializable
data class DeckSlotConfig(
    val slot: Int,
    val actionType: String,
    val settings: Map<String, JsonElement>? = null,
)

// --- Multi-session discovery ---

@Serializable
data class SessionInfo(
    val id: String,
    val port: Int,
    val projectName: String? = null,
    val agentType: String? = null,
    val alive: Boolean = true,
    val state: String? = null,
)

// --- Bridge timeline entry (rich OpenClaw events) ---

@Serializable
data class BridgeTimelineEntry(
    val ts: Long,
    val type: String,
    val raw: String,
    val detail: String? = null,
    val approvalId: String? = null,
    val status: String? = null,
    val agentType: String? = null,
)

fun BridgeTimelineEntry.toTimelineEntry() = dev.agentdeck.state.TimelineEntry(
    timestamp = ts,
    type = type,
    summary = raw,
    detail = detail,
    agentType = agentType,
    status = status,
)

sealed class BridgeEvent {
    data class State(val data: StateUpdate) : BridgeEvent()
    data class Usage(val data: UsageUpdate) : BridgeEvent()
    data class Voice(val data: VoiceState) : BridgeEvent()
    data class Connected(val sessionId: String?) : BridgeEvent()
    data object Disconnected : BridgeEvent()
    data class DisplaySleep(val displayOn: Boolean) : BridgeEvent()
    data class SessionsList(val sessions: List<SessionInfo>) : BridgeEvent()
    data class EncoderState(val encoders: List<EncoderSlotState>, val takeoverActive: Boolean) : BridgeEvent()
    data class ButtonState(val buttons: List<ButtonSlotState>) : BridgeEvent()
    data class SlotMap(val buttons: List<DeckSlotConfig>, val encoders: List<DeckSlotConfig>) : BridgeEvent()
    data class Timeline(val entry: BridgeTimelineEntry, val upsert: Boolean = false) : BridgeEvent()
    data class TimelineHistory(val entries: List<BridgeTimelineEntry>) : BridgeEvent()
    data class UserPrompt(val text: String) : BridgeEvent()
}

// --- App -> Bridge commands ---

object PluginCommands {
    fun respond(value: String): String =
        """{"type":"respond","value":${Json.encodeToString(kotlinx.serialization.serializer<String>(), value)}}"""

    fun selectOption(index: Int): String =
        """{"type":"select_option","index":$index}"""

    fun sendPrompt(text: String): String =
        """{"type":"send_prompt","text":${Json.encodeToString(kotlinx.serialization.serializer<String>(), text)}}"""

    fun interrupt(): String = """{"type":"interrupt"}"""

    fun escape(): String = """{"type":"escape"}"""

    fun queryUsage(): String = """{"type":"query_usage"}"""

    fun switchMode(): String = """{"type":"switch_mode"}"""

    fun utility(action: String, value: Int? = null): String {
        val valueStr = if (value != null) ""","value":$value""" else ""
        return """{"type":"utility","action":"$action"$valueStr}"""
    }

    fun navigateOption(direction: String): String =
        """{"type":"navigate_option","direction":"$direction"}"""
}

// --- JSON parsing ---

val protocolJson = Json {
    ignoreUnknownKeys = true
    isLenient = true
    coerceInputValues = true
}

fun parseBridgeMessage(text: String): BridgeEvent? {
    return try {
        val element = protocolJson.parseToJsonElement(text)
        val obj = element.jsonObject
        val type = obj["type"]?.jsonPrimitive?.content ?: return null

        when (type) {
            "state_update" -> {
                val data = protocolJson.decodeFromJsonElement<StateUpdate>(element)
                BridgeEvent.State(data)
            }
            "usage_update" -> {
                val data = protocolJson.decodeFromJsonElement<UsageUpdate>(element)
                BridgeEvent.Usage(data)
            }
            "voice_state" -> {
                val data = protocolJson.decodeFromJsonElement<VoiceState>(element)
                BridgeEvent.Voice(data)
            }
            "connection" -> {
                val status = obj["status"]?.jsonPrimitive?.content
                val sessionId = obj["sessionId"]?.jsonPrimitive?.content
                if (status == "connected") BridgeEvent.Connected(sessionId)
                else BridgeEvent.Disconnected
            }
            "display_state" -> {
                val displayOn = obj["displayOn"]?.jsonPrimitive?.boolean ?: true
                BridgeEvent.DisplaySleep(displayOn)
            }
            "sessions_list" -> {
                val sessionsArray = obj["sessions"]
                if (sessionsArray != null) {
                    val sessions = protocolJson.decodeFromJsonElement<List<SessionInfo>>(sessionsArray)
                    BridgeEvent.SessionsList(sessions)
                } else null
            }
            "encoder_state" -> {
                val encodersArray = obj["encoders"]
                val takeoverActive = obj["takeoverActive"]?.jsonPrimitive?.boolean ?: false
                if (encodersArray != null) {
                    val encoders = protocolJson.decodeFromJsonElement<List<EncoderSlotState>>(encodersArray)
                    BridgeEvent.EncoderState(encoders, takeoverActive)
                } else null
            }
            "button_state" -> {
                val buttonsArray = obj["buttons"]
                if (buttonsArray != null) {
                    val buttons = protocolJson.decodeFromJsonElement<List<ButtonSlotState>>(buttonsArray)
                    BridgeEvent.ButtonState(buttons)
                } else null
            }
            "deck_slot_map" -> {
                val buttonsArray = obj["buttons"]
                val encodersArray = obj["encoders"]
                if (buttonsArray != null && encodersArray != null) {
                    val buttons = protocolJson.decodeFromJsonElement<List<DeckSlotConfig>>(buttonsArray)
                    val encoders = protocolJson.decodeFromJsonElement<List<DeckSlotConfig>>(encodersArray)
                    BridgeEvent.SlotMap(buttons, encoders)
                } else null
            }
            "user_prompt" -> {
                val promptText = obj["text"]?.jsonPrimitive?.content
                if (promptText != null) BridgeEvent.UserPrompt(promptText) else null
            }
            "timeline_event" -> {
                val entryObj = obj["entry"]
                val isUpsert = obj["upsert"]?.jsonPrimitive?.boolean ?: false
                if (entryObj != null) {
                    val entry = protocolJson.decodeFromJsonElement<BridgeTimelineEntry>(entryObj)
                    BridgeEvent.Timeline(entry, isUpsert)
                } else null
            }
            "timeline_history" -> {
                val entriesArray = obj["entries"]
                if (entriesArray != null) {
                    val entries = protocolJson.decodeFromJsonElement<List<BridgeTimelineEntry>>(entriesArray)
                    BridgeEvent.TimelineHistory(entries)
                } else null
            }
            else -> null
        }
    } catch (e: Exception) {
        Log.e("Terrarium", "parseBridgeMessage failed: ${e.message}, raw=${text.take(300)}")
        null
    }
}
